AK/Noncopyable.h

```diff
@@ -25,3 +25,44 @@ public:                            \
 public:                             \
     c(c const&) = default;          \
     c& operator=(c const&) = default
+
+#define AK_MAKE_CONDITIONALLY_NONMOVABLE(c, ...)              \
+public:                                                       \
+    c(c&&)                                                    \
+    requires(!(AK::Detail::IsMoveConstructible __VA_ARGS__))  \
+    = delete;                                                 \
+    c& operator=(c&&)                                         \
+    requires(!(AK::Detail::IsMoveConstructible __VA_ARGS__)   \
+                || !(AK::Detail::IsDestructible __VA_ARGS__)) \
+    = delete
+
+#define AK_MAKE_CONDITIONALLY_MOVABLE(c, T) \
+    AK_MAKE_CONDITIONALLY_NONMOVABLE(c, T); \
+    c(c&&) = default;                       \
+    c& operator=(c&&) = default
+
+#define AK_MAKE_CONDITIONALLY_NONCOPYABLE(c, ...)             \
+public:                                                       \
+    c(c const&)                                               \
+    requires(!(AK::Detail::IsCopyConstructible __VA_ARGS__))  \
+    = delete;                                                 \
+    c& operator=(c const&)                                    \
+    requires(!(AK::Detail::IsCopyConstructible __VA_ARGS__)   \
+                || !(AK::Detail::IsDestructible __VA_ARGS__)) \
+    = delete
+
+#define AK_MAKE_CONDITIONALLY_COPYABLE(c, ...)         \
+    AK_MAKE_CONDITIONALLY_NONCOPYABLE(c, __VA_ARGS__); \
+    c(c const&) = default;                             \
+    c& operator=(c const&) = default
+
+#define AK_MAKE_CONDITIONALLY_NONDESTRUCTIBLE(c, ...)   \
+    ~c()                                                \
+    requires(!(AK::Detail::IsDestructible __VA_ARGS__)) \
+    = delete
+
+#define AK_MAKE_CONDITIONALLY_DESTRUCTIBLE(c, ...)      \
+    ~c()                                                \
+    requires(!(AK::Detail::IsDestructible __VA_ARGS__)) \
+    = delete;                                           \
```

Comment:
I do not like that these do not take care of `<>` and type packs on their own, but I can see that getting those to work might be a bit more involved